lnwire/dyn_propose.go

```diff
@@ -76,11 +76,19 @@ func (dp *DynPropose) Encode(w *bytes.Buffer, _ uint32) error {
 
 	producers := dynProposeRecords(dp)
 
-	err := EncodeMessageExtraData(&dp.ExtraData, producers...)
+	// Encode all known records.
+	var tlvData ExtraOpaqueData
+	err := tlvData.PackRecords(producers...)
 	if err != nil {
 		return err
 	}
 
+	// Write the known records.
+	if err := WriteBytes(w, tlvData); err != nil {
+		return err
+	}
+
+	// Encode ExtraData.
 	return WriteBytes(w, dp.ExtraData)
```

Comment:
i think it was correct before. all TLV data should be encoded together because the ordering matters. With this change, what can happen is we have producers for type 0, 1, 4, 5 but not for type 3 - instead type 3 is an unknown type for us so we just have it in `ExtraData`. WIth this change, it would be encoded in the order: 0, 1, 4, 5, 3 which is invalid TLV. 

So all TLV (unknown types and our known types) should all be packed into the same TLV stream so that order is correct

Reply:
> The previous was wrong, it would decode the raw bytes correctly, but when encoding it would always put known tlv bytes into the extra byte,

I dont think this is wrong. Yes the `lnwire.*` message will have both fields populated - but that doesnt matter. What matters is that at encoding time we just done write the values twice (which we dont). 

If you want to keep the in-memory struct more lean, then i suggest us reviving this [this PR](https://github.com/lightningnetwork/lnd/pull/9175/commits/344e06a57c94bc34441ded3c3d0ac309f846b37f) see here how we then only keep the _signed_ extra fields in the ExtraBytes field and then at encoding time, we merge them all togther so that the ordering is correct. That way we arent carrying around duplicate data internally

Reply:
hmm I think it would be easier if you take a look at `TestDynAckEncodeDecode` or any of the new tests which have the new hard code bytes and let me know what went off. **Unless the wire bytes** used are wrong, I don't think any of the encoding/decoding is wrong - it correctly handles the case you mentioned.

> Yes but that is only for our known TLVs that we have producers for.

We use `da.ExtraData.RecordProducers()` in a latter commit.

> I dont think this is wrong. Yes the lnwire.* message will have both fields populated - but that doesnt matter. What matters is that at encoding time we just done write the values twice (which we dont).

We did, as shown in the test. The raw bytes will be decoded into a msg, then encoded into bytes, but they don't match.